src/event.cpp

```diff
@@ -1,38 +1,111 @@
-#pragma once
-
+#include <ctime>
+#include <openssl/evp.h>
+#include <iomanip>
+#include <sstream>
 #include <string>
-#include <vector>
 #include <nlohmann/json.hpp>
+#include <sha.h>
 
 #include "nostr.hpp"
 
+using nlohmann::json;
+using std::hex;
+using std::invalid_argument;
+using std::make_shared;
+using std::setw;
+using std::setfill;
+using std::shared_ptr;
 using std::string;
+using std::stringstream;
+using std::time;
 
 namespace nostr 
 {
-    nlohmann::json Event::serialize() const
+string Event::serialize()
+{
+    try
     {
-        nlohmann::json j = {
-            {"id", this->id},
-            {"pubkey", this->pubkey},
-            {"created_at", this->created_at},
-            {"kind", this->kind},
-            {"tags", this->tags},
-            {"content", this->content},
-            {"sig", this->sig}
-        };
-        return j.dump();
+        this->validate();
+    }
+    catch (const invalid_argument& e)
+    {
+        throw e;
+    }
+
+    json j = {
+        {"pubkey", this->pubkey},
+        {"created_at", this->createdAt},
+        {"kind", this->kind},
+        {"tags", this->tags},
+        {"content", this->content},
+        {"sig", this->sig}
     };
 
-    void Event::deserialize(string jsonString)
+    j["id"] = this->generateId(j.dump());
+
+    return j.dump();
+};
+
+Event Event::fromString(string jstr)
+{
+    json j = json::parse(jstr);
+
+    return Event::fromJson(j);
+};
+
+Event Event::fromJson(json j)
+{
+    Event event;
+
+    event.id = j["id"];
```

Comment:
Unchecked access with [] operator is often a source of failure. Better to do check access with default or add some try catch. ( https://json.nlohmann.me/features/element_access/ )